TestIcebergSparkCompatibility.testRegisterTableWithShowCreateTable flaky

```
2024-03-28 15:16:37 INFO: FAILURE     /    io.trino.tests.product.iceberg.TestIcebergSparkCompatibility.testRegisterTableWithShowCreateTable [AVRO] (Groups: profile_specific_tests, iceberg) took 3.1 seconds
tests               | 2024-03-28 15:16:37 SEVERE: Failure cause:
tests               | io.trino.tempto.query.QueryExecutionException: java.sql.SQLException: Query failed (#20240328_093137_00698_eqax2): More than one latest metadata file found at location: hdfs://hadoop-master:9000/user/hive/warehouse/test_register_table_with_show_create_table_avro_4pej3h6lg5/metadata, latest metadata files are [hdfs://hadoop-master:9000/user/hive/warehouse/test_register_table_with_show_create_table_avro_4pej3h6lg5/metadata/00002-d126c010-497c-4820-a1b7-503fe8e77b05.metadata.json, hdfs://hadoop-master:9000/user/hive/warehouse/test_register_table_with_show_create_table_avro_4pej3h6lg5/metadata/00002-dc5c8fa5-67a8-42d4-8675-87c57c83c726.metadata.json]
tests               | 	at io.trino.tempto.query.JdbcQueryExecutor.execute(JdbcQueryExecutor.java:119)
tests               | 	at io.trino.tempto.query.JdbcQueryExecutor.executeQuery(JdbcQueryExecutor.java:84)
tests               | 	at io.trino.tests.product.utils.QueryExecutors$1.lambda$executeQuery$0(QueryExecutors.java:54)
tests               | 	at dev.failsafe.Functions.lambda$toCtxSupplier$11(Functions.java:243)
tests               | 	at dev.failsafe.Functions.lambda$get$0(Functions.java:46)
tests               | 	at dev.failsafe.internal.RetryPolicyExecutor.lambda$apply$0(RetryPolicyExecutor.java:74)
tests               | 	at dev.failsafe.SyncExecutionImpl.executeSync(SyncExecutionImpl.java:187)
tests               | 	at dev.failsafe.FailsafeExecutor.call(FailsafeExecutor.java:376)
tests               | 	at dev.failsafe.FailsafeExecutor.get(FailsafeExecutor.java:112)
tests               | 	at io.trino.tests.product.utils.QueryExecutors$1.executeQuery(QueryExecutors.java:54)
tests               | 	at io.trino.tests.product.iceberg.TestIcebergSparkCompatibility.testRegisterTableWithShowCreateTable(TestIcebergSparkCompatibility.java:2592)
tests               | 	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
tests               | 	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
tests               | 	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:104)
tests               | 	at org.testng.internal.Invoker.invokeMethod(Invoker.java:645)
tests               | 	at org.testng.internal.Invoker.invokeTestMethod(Invoker.java:851)
tests               | 	at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:1177)
tests               | 	at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:129)
tests               | 	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:112)
tests               | 	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
tests               | 	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
tests               | 	at java.base/java.lang.Thread.run(Thread.java:1583)
tests               | Caused by: java.sql.SQLException: Query failed (#20240328_093137_00698_eqax2): More than one latest metadata file found at location: hdfs://hadoop-master:9000/user/hive/warehouse/test_register_table_with_show_create_table_avro_4pej3h6lg5/metadata, latest metadata files are [hdfs://hadoop-master:9000/user/hive/warehouse/test_register_table_with_show_create_table_avro_4pej3h6lg5/metadata/00002-d126c010-497c-4820-a1b7-503fe8e77b05.metadata.json, hdfs://hadoop-master:9000/user/hive/warehouse/test_register_table_with_show_create_table_avro_4pej3h6lg5/metadata/00002-dc5c8fa5-67a8-42d4-8675-87c57c83c726.metadata.json]
tests               | 	at io.trino.jdbc.AbstractTrinoResultSet.resultsException(AbstractTrinoResultSet.java:1937)
tests               | 	at io.trino.jdbc.TrinoResultSet$ResultsPageIterator.computeNext(TrinoResultSet.java:294)
tests               | 	at io.trino.jdbc.TrinoResultSet$ResultsPageIterator.computeNext(TrinoResultSet.java:254)
tests               | 	at io.trino.jdbc.$internal.guava.collect.AbstractIterator.tryToComputeNext(AbstractIterator.java:145)
tests               | 	at io.trino.jdbc.$internal.guava.collect.AbstractIterator.hasNext(AbstractIterator.java:140)
tests               | 	at java.base/java.util.Spliterators$IteratorSpliterator.tryAdvance(Spliterators.java:1949)
tests               | 	at java.base/java.util.stream.StreamSpliterators$WrappingSpliterator.lambda$initPartialTraversalState$0(StreamSpliterators.java:292)
tests               | 	at java.base/java.util.stream.StreamSpliterators$AbstractWrappingSpliterator.fillBuffer(StreamSpliterators.java:206)
tests               | 	at java.base/java.util.stream.StreamSpliterators$AbstractWrappingSpliterator.doAdvance(StreamSpliterators.java:161)
tests               | 	at java.base/java.util.stream.StreamSpliterators$WrappingSpliterator.tryAdvance(StreamSpliterators.java:298)
tests               | 	at java.base/java.util.Spliterators$1Adapter.hasNext(Spliterators.java:681)
tests               | 	at io.trino.jdbc.TrinoResultSet$AsyncIterator.lambda$new$1(TrinoResultSet.java:179)
tests               | 	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:572)
tests               | 	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317)
tests               | 	... 3 more
tests               | 	Suppressed: java.lang.Exception: Query: CALL iceberg.system.register_table ('default', 'test_register_table_with_show_create_table_avro_4pej3h6lg5', 'hdfs://hadoop-master:9000/user/hive/warehouse/test_register_table_with_show_create_table_avro_4pej3h6lg5')
tests               | 		at io.trino.tempto.query.JdbcQueryExecutor.executeQueryNoParams(JdbcQueryExecutor.java:136)
tests               | 		at io.trino.tempto.query.JdbcQueryExecutor.execute(JdbcQueryExecutor.java:112)
tests               | 		at io.trino.tempto.query.JdbcQueryExecutor.executeQuery(JdbcQueryExecutor.java:84)
tests               | 		at io.trino.tests.product.utils.QueryExecutors$1.lambda$executeQuery$0(QueryExecutors.java:54)
tests               | 		at dev.failsafe.Functions.lambda$toCtxSupplier$11(Functions.java:243)
tests               | 		at dev.failsafe.Functions.lambda$get$0(Functions.java:46)
tests               | 		at dev.failsafe.internal.RetryPolicyExecutor.lambda$apply$0(RetryPolicyExecutor.java:74)
tests               | 		at dev.failsafe.SyncExecutionImpl.executeSync(SyncExecutionImpl.java:187)
tests               | 		at dev.failsafe.FailsafeExecutor.call(FailsafeExecutor.java:376)
tests               | 		at dev.failsafe.FailsafeExecutor.get(FailsafeExecutor.java:112)
tests               | 		at io.trino.tests.product.utils.QueryExecutors$1.executeQuery(QueryExecutors.java:54)
tests               | 		at io.trino.tests.product.iceberg.TestIcebergSparkCompatibility.testRegisterTableWithShowCreateTable(TestIcebergSparkCompatibility.java:2592)
tests               | 		at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
tests               | 		at java.base/java.lang.reflect.Method.invoke(Method.java:580)
tests               | 		at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:104)
tests               | 		at org.testng.internal.Invoker.invokeMethod(Invoker.java:645)
tests               | 		at org.testng.internal.Invoker.invokeTestMethod(Invoker.java:851)
tests               | 		at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:1177)
tests               | 		at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:129)
tests               | 		at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:112)
tests               | 		... 3 more
tests               | Caused by: io.trino.spi.TrinoException: More than one latest metadata file found at location: hdfs://hadoop-master:9000/user/hive/warehouse/test_register_table_with_show_create_table_avro_4pej3h6lg5/metadata, latest metadata files are [hdfs://hadoop-master:9000/user/hive/warehouse/test_register_table_with_show_create_table_avro_4pej3h6lg5/metadata/00002-d126c010-497c-4820-a1b7-503fe8e77b05.metadata.json, hdfs://hadoop-master:9000/user/hive/warehouse/test_register_table_with_show_create_table_avro_4pej3h6lg5/metadata/00002-dc5c8fa5-67a8-42d4-8675-87c57c83c726.metadata.json]
tests               | 	at io.trino.plugin.iceberg.IcebergUtil.getLatestMetadataLocation(IcebergUtil.java:895)
tests               | 	at io.trino.plugin.iceberg.procedure.RegisterTableProcedure.lambda$getMetadataLocation$1(RegisterTableProcedure.java:174)
tests               | 	at java.base/java.util.Optional.orElseGet(Optional.java:364)
tests               | 	at io.trino.plugin.iceberg.procedure.RegisterTableProcedure.getMetadataLocation(RegisterTableProcedure.java:174)
tests               | 	at io.trino.plugin.iceberg.procedure.RegisterTableProcedure.doRegisterTable(RegisterTableProcedure.java:139)
tests               | 	at io.trino.plugin.iceberg.procedure.RegisterTableProcedure.registerTable(RegisterTableProcedure.java:108)
tests               | 	at java.base/java.lang.invoke.MethodHandle.invokeWithArguments(MethodHandle.java:733)
tests               | 	at java.base/java.lang.invoke.MethodHandle.invokeWithArguments(MethodHandle.java:758)
tests               | 	at io.trino.execution.CallTask.execute(CallTask.java:210)
tests               | 	at io.trino.execution.CallTask.execute(CallTask.java:70)
tests               | 	at io.trino.execution.DataDefinitionExecution.start(DataDefinitionExecution.java:146)
tests               | 	at io.trino.execution.SqlQueryManager.createQuery(SqlQueryManager.java:272)
tests               | 	at io.trino.dispatcher.LocalDispatchQuery.startExecution(LocalDispatchQuery.java:145)
tests               | 	at io.trino.dispatcher.LocalDispatchQuery.lambda$waitForMinimumWorkers$2(LocalDispatchQuery.java:129)
tests               | 	at io.airlift.concurrent.MoreFutures.lambda$addSuccessCallback$12(MoreFutures.java:570)
tests               | 	at io.airlift.concurrent.MoreFutures$3.onSuccess(MoreFutures.java:545)
tests               | 	at com.google.common.util.concurrent.Futures$CallbackListener.run(Futures.java:1137)
tests               | 	at io.trino.$gen.Trino_76c956e____20240328_092351_2.run(Unknown Source)
tests               | 	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
tests               | 	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
tests               | 	at java.base/java.lang.Thread.run(Thread.java:1583)
tests               | 
```

cc: @findinpath

Input from @pajaks 

Logs: https://github.com/starburstdata/cork/actions/runs/8464874870/job/23190847390

```
2024-03-28T09:31:36.0062329Z presto-master       | Caused by: org.apache.hadoop.ipc.RemoteException(java.io.IOException): File /user/hive/warehouse/test_register_table_with_show_create_table_avro_4pej3h6lg5/metadata/00002-dc5c8fa5-67a8-42d4-8675-87c57c83c726.metadata.json could only be written to 0 of the 1 minReplication nodes. There are 1 datanode(s) running and no node(s) are excluded in this operation.
```

Looks like during retry second metadata file is created.

--- a/testing/trino-product-tests/src/main/java/io/trino/tests/product/iceberg/TestIcebergSparkCompatibility.java
+++ b/testing/trino-product-tests/src/main/java/io/trino/tests/product/iceberg/TestIcebergSparkCompatibility.java
@@ -67,6 +67,7 @@ import static io.trino.tests.product.TestGroups.PROFILE_SPECIFIC_TESTS;
 import static io.trino.tests.product.iceberg.TestIcebergSparkCompatibility.CreateMode.CREATE_TABLE_AND_INSERT;
 import static io.trino.tests.product.iceberg.TestIcebergSparkCompatibility.CreateMode.CREATE_TABLE_AS_SELECT;
 import static io.trino.tests.product.iceberg.TestIcebergSparkCompatibility.CreateMode.CREATE_TABLE_WITH_NO_DATA_AND_INSERT;
+import static io.trino.tests.product.iceberg.util.IcebergTestUtils.getLatestMetadataFilename;
 import static io.trino.tests.product.iceberg.util.IcebergTestUtils.getTableLocation;
 import static io.trino.tests.product.iceberg.util.IcebergTestUtils.stripNamenodeURI;
 import static io.trino.tests.product.utils.QueryExecutors.onHive;
@@ -2524,9 +2525,8 @@ public class TestIcebergSparkCompatibility
 
         onSpark().executeQuery(format("CREATE TABLE %s (a INT, b STRING, c BOOLEAN) USING ICEBERG TBLPROPERTIES ('write.format.default' = '%s')", sparkTableName, storageFormat));
         onSpark().executeQuery(format("INSERT INTO %s values(1, 'INDIA', true)", sparkTableName));
-        onTrino().executeQuery(format("INSERT INTO %s values(2, 'USA', false)", trinoTableName));
 
-        List<Row> expected = List.of(row(1, "INDIA", true), row(2, "USA", false));
+        List<Row> expected = List.of(row(1, "INDIA", true));
         String tableLocation = getTableLocation(trinoTableName);
         // Drop table from hive metastore and use the same table name to register again with the metadata
         dropTableFromMetastore(baseTableName);
@@ -2555,10 +2555,11 @@ public class TestIcebergSparkCompatibility
         onTrino().executeQuery(format("COMMENT ON COLUMN %s.c is 'c-comment'", trinoTableName));
 
         String tableLocation = getTableLocation(trinoTableName);
+        String metadataFileName = getLatestMetadataFilename(TRINO_CATALOG, TEST_SCHEMA_NAME, baseTableName);
         // Drop table from hive metastore and use the same table name to register again with the metadata
         dropTableFromMetastore(baseTableName);
 
-        onTrino().executeQuery(format("CALL iceberg.system.register_table ('%s', '%s', '%s')", TEST_SCHEMA_NAME, baseTableName, tableLocation));
+        onTrino().executeQuery(format("CALL iceberg.system.register_table ('%s', '%s', '%s', '%s')", TEST_SCHEMA_NAME, baseTableName, tableLocation, metadataFileName));
 
         assertThat(getTableComment(baseTableName)).isEqualTo("my-table-comment");
         assertThat(getColumnComment(baseTableName, "a")).isEqualTo("a-comment");
@@ -2586,10 +2587,11 @@ public class TestIcebergSparkCompatibility
         List<Row> expectedShowCreateTableRows = expectedShowCreateTable.rows().stream().map(columns -> row(columns.toArray())).collect(toImmutableList());
 
         String tableLocation = getTableLocation(trinoTableName);
+        String metadataFileName = getLatestMetadataFilename(TRINO_CATALOG, TEST_SCHEMA_NAME, baseTableName);
         // Drop table from hive metastore and use the same table name to register again with the metadata
         dropTableFromMetastore(baseTableName);
 
-        onTrino().executeQuery(format("CALL iceberg.system.register_table ('%s', '%s', '%s')", TEST_SCHEMA_NAME, baseTableName, tableLocation));
+        onTrino().executeQuery(format("CALL iceberg.system.register_table ('%s', '%s', '%s', '%s')", TEST_SCHEMA_NAME, baseTableName, tableLocation, metadataFileName));
 
         QueryResult actualDescribeTable = onSpark().executeQuery("DESCRIBE TABLE EXTENDED " + sparkTableName);
         QueryResult actualShowCreateTable = onTrino().executeQuery("SHOW CREATE TABLE " + trinoTableName);
@@ -2612,10 +2614,11 @@ public class TestIcebergSparkCompatibility
         onTrino().executeQuery(format("INSERT INTO %s values(2, 'USA', false)", trinoTableName));
 
         String tableLocation = getTableLocation(trinoTableName);
+        String metadataFileName = getLatestMetadataFilename(TRINO_CATALOG, TEST_SCHEMA_NAME, baseTableName);
         // Drop table from hive metastore and use the same table name to register again with the metadata
         dropTableFromMetastore(baseTableName);
 
-        onTrino().executeQuery(format("CALL iceberg.system.register_table ('%s', '%s', '%s')", TEST_SCHEMA_NAME, baseTableName, tableLocation));
+        onTrino().executeQuery(format("CALL iceberg.system.register_table ('%s', '%s', '%s', '%s')", TEST_SCHEMA_NAME, baseTableName, tableLocation, metadataFileName));
         onSpark().executeQuery(format("INSERT INTO %s values(3, 'POLAND', true)", sparkTableName));
 
         List<Row> expected = List.of(row(1, "INDIA", true), row(2, "USA", false), row(3, "POLAND", true));
@@ -2659,13 +2662,14 @@ public class TestIcebergSparkCompatibility
         onTrino().executeQuery(format("INSERT INTO %s values(2, 'USA', false)", trinoTableName));
 
         String tableLocation = getTableLocation(trinoTableName);
+        String metadataFileName = getLatestMetadataFilename(TRINO_CATALOG, TEST_SCHEMA_NAME, baseTableName);
         String baseTableNameNew = baseTableName + "_new";
         String trinoTableNameNew = trinoTableName(baseTableNameNew);
         String sparkTableNameNew = sparkTableName(baseTableNameNew);
         // Drop table from hive metastore and use the same table name to register again with the metadata
         dropTableFromMetastore(baseTableName);
 
-        onTrino().executeQuery(format("CALL iceberg.system.register_table ('%s', '%s', '%s')", TEST_SCHEMA_NAME, baseTableNameNew, tableLocation));
+        onTrino().executeQuery(format("CALL iceberg.system.register_table ('%s', '%s', '%s', '%s')", TEST_SCHEMA_NAME, baseTableNameNew, tableLocation, metadataFileName));
         onSpark().executeQuery(format("INSERT INTO %s values(3, 'POLAND', true)", sparkTableNameNew));
         List<Row> expected = List.of(row(1, "INDIA", true), row(2, "USA", false), row(3, "POLAND", true));
 

--- a/testing/trino-product-tests/src/main/java/io/trino/tests/product/iceberg/util/IcebergTestUtils.java
+++ b/testing/trino-product-tests/src/main/java/io/trino/tests/product/iceberg/util/IcebergTestUtils.java
@@ -26,6 +26,11 @@ public final class IcebergTestUtils
         return (String) onTrino().executeQuery("SELECT DISTINCT regexp_replace(\"$path\", '/[^/]*/[^/]*$', '') FROM " + tableName).getOnlyValue();
     }
 
+    public static String getLatestMetadataFilename(String catalog, String schema, String tableName)
+    {
+        return (String) onTrino().executeQuery("SELECT substring(file, strpos(file, '/', -1) + 1) FROM %s.%s.\"%s$metadata_log_entries\" ORDER BY timestamp DESC LIMIT 1".formatted(catalog, schema, tableName)).getOnlyValue();
+    }
+
     public static String stripNamenodeURI(String location)
     {
         return URI.create(location).getPath();

